### DISCUSSION TITLE
Flows: Show snackbars, alert dialogs, confirm dialogs, and prompt dialogs to the user

### DISCUSSION THREAD

**Discussion by silasabbott** [directus/directus#21163] [Feature Requests]
### Summary

Allow Flows to send snackbars and/or popup dialogs to notify the user of things happening at any stage in the Flow. Dialogs could potentially cancel the Flow, or even accept data from the user to use in the Flow.

### Motivation

Let's say I'm using Directus for a blog. When a user creates a new article and saves it, a flow is triggered that posts the article to Facebook and Instagram over an external API. Currently, there's no way to clearly notify the user whether this (or any other flow) succeeded or failed.

Furthermore, the user *should* be able to confirm or cancel a Flow if it's about to perform a destructive or high-impact action that the user triggered. A confirmation dialog could prevent serious unwanted actions from occurring if the user didn't mean to trigger something or has changed their mind.

At the very least, dialogs could show an example of or explain something that a Flow is about to do before it happens—ensuring that the user is aware of what's happening.

I purposefully mentioned both snackbars and dialogs because they both serve different purposes: 
- Snackbars are an unobtrusive way to notify the user of small, unimportant changes so they aren't left wondering if something completed or failed
- Dialogs are a direct and upfront way to ensure a user sees important information by forcing some sort of interaction before continuing

### Detailed Design

Here are the exact changes I propose:
1. **Add Snackbars to Directus that show when a new Notification arrives.** Notifications are already implemented in Directus, and Flows can send them. Instead of adding a whole new Operation for this, I suggest simply showing a snackbar whenever a new Notification arrives while a user is logged in.
2. **Create a new "Dialog" Flow Operation.** Adding a new Dialog Operation allows Flows to be paused until the user interacts. When adding a Dialog to a Flow, it could have 3 options for the type of Dialog to show:
    - **Alert:** A dialog with a title, message, and a single "OK" button
    - **Confirm:** A dialog with a title, message, and "Cancel" and "Confirm" buttons. This dialog type would have the ability to cancel the current Flow from executing further flow operations.
    - **Prompt:** A dialog with a title, message, a single field for the user to input data, and a "Cancel" and "Confirm". The input field could be any basic Directus field type, passed on to the next flow operation

### Must Have:
- **Snackbars:**
    - Send a snackbar to the current user from a Flow
    - Use data from the previous Flow operation in the snackbar message
- **Dialogs:**
    - Show a dialog to the current user from a Flow
    - Use data from the previous Flow operation in the dialog
    - All dialogs must prevent further interaction until dismissed
    - All dialogs must block Flow steps until the confirmation button is pressed
    - An `Alert` type of dialog that has a title, message, and single "OK" (confirm) button

### Should Have:
- **Dialogs:**
    - A `Confirm` type of dialog that has a title, message, and "Cancel" and "Confirm" buttons. Flows should cancel and not execute any more steps if the "Cancel" button is pressed

### Could Have:
- **Snackbars:**
    - Send snackbars to all users currently logged in to Directus
    - Add a link to the snackbar so the user could click it (for example: "Your video was published to all channels! Click to view live statistics →")
- **Dialogs:**
    - A `Prompt` type of dialog that has a title, message, data input field, and "Cancel" and "Confirm" buttons
    - Customize the "Cancel" and/or "Confirm" button text on dialogs
    - Choose any type of data input field for prompt dialogs
### Won't Have:
- To keep things simple, we shouldn't allow more than one data input field for prompt dialogs
- Advanced formatting (tables, images, etc.) in dialog messages are probably not necessary at this time

### Drawbacks

- If a Flow is not triggered by a user action (or a Dialog is shown later in the Flow) Dialogs could be surprising or unexpected and cause the user to accidentally confirm or cancel an action. This could maybe be mitigated by showing an overlay with a loading indicator for all other steps in a Flow that contains a Dialog Operation
- Snackbars could potentially cover UI elements

### Alternatives

I'm not aware of any alternative solutions to this, but would love to hear any more ideas!

If this is not implemented, it could continue to allow serious issues with certain Flows: 
- Flows handling sensitive data could share that data or take action with that data without the ability to notify or gain consent from the Directus user
- Flows with destructive actions could execute without notifying or getting confirmation from the Directus user
- Flows that might need additional info from the user during execution will not be possible
- After performing an action that triggers a Flow, a Directus user will have no way of knowing if a flow operation succeeded or failed

### Adoption Strategy

This won't be a breaking change and will mostly be opt-in. Snackbars showing whenever a Notification arrives will be the only potentially jarring change, and it could be enabled or disabled in settings.

### Unresolved Questions

- Should snackbars have their own Flow Operations, or should they simply be shown whenever a new Notification arrives?
- Should all three types of dialogs be implemented (alert, confirm, and prompt)?

**Comment by islam-str** [comment]
+1

**Comment by MichalNemec** [comment]
This is very needed if there are multiple people involved in data management. Why nobody talks about it?

**Comment by pepeday** [comment]
I support this. At a minimum, a customizable snack bar as feedback for the user running a manual trigger is definitely needed.

**Comment by Borty97** [comment]
This is very interesting. I'm looking for some extension (operation) to do this but there is no one!

**Comment by IamSAB** [comment]
Hey, just stumled accross this because I was looking for the exact feature.

In my case, I would like to add a trigger to create invitations for an event based on selected users. That is actually a potential workaround to easy add M2M pivot values. Instead of create each junction entry by yourself, you only select desired users and the flow creates the junction entry with the invited status.
